File: lib/board.rb
class Board
  attr_reader :width, :height

  def initialize(width, height)
    @width = width
    @height = height
    @state = Array.new(width) { Array.new(height, 0) }
  end

  def scan(x, y, player_id)
    @state[x][y] = player_id
  end

  # Returns string representation of board state
  def draw
    str = ''

    @state.each do |row|
      str << '|'

      row.each do |val|
        str << val.to_s
        str << '|'
      end

      str << "\n"
    end

    str
  end
end

File: lib/board.js
class Board {
  constructor(width, height) {
    this.width = width;
    this.height = height;
    this.state = this.create2DArray(width, height);
  }

  scan(x, y, playerId) {
    this.state[x][y] = playerId;
  }

  // Returns a string representation of the board state
  draw() {
    let str = ''

    this.state.forEach(row => {
      str += '|'

      row.forEach(val => {
        str += val;
        str += '|';
      });

      str += '\n';
    });

    return str;
  }

  asJSON() {
    return {
      tiles: this.state,
      width: this.width,
      height: this.height
    };
  }

  create2DArray(width, height) {
    let rows = new Array(width);
    for (let i = 0; i < width; i++) {
      rows[i] = new Array(height).fill(0);
    }

    return rows;
  }
}

module.exports = Board;
